site/_posts/2018-12-12-reinventing-the-proxy.md

```diff
@@ -0,0 +1,124 @@
+
+Remote desktop consoles like [VNC](https://en.wikipedia.org/wiki/Virtual_Network_Computing), [SPICE](https://en.wikipedia.org/wiki/Simple_Protocol_for_Independent_Computing_Environments) or [RDP](https://en.wikipedia.org/wiki/Remote_Desktop_Protocol) are using bidirectional TCP streams for data transfer. On the other hand, HTTP is a request-response protocol with a much shorter connection lifetime. However, thanks to the [WebSocket](https://en.wikipedia.org/wiki/WebSocket) standard it is possible to upgrade a HTTP request to a bidirectional stream. By using a [WebSocket-to-TCP proxy](https://github.com/novnc/websockify), it is possible to forward a remote console connection to a web browser. If the browser has access to a JavaScript library that implements the chosen protocol, it is possible to render a remote screen into an HTML5 \<canvas\> element.
+
+Our architecture is similar to the one described above. The proxy process was written in Python and it was forked from Ruby each connection on a separate TCP port. This port had to be opened on the firewall, so the browser could connect to it directly. It was not an ideal solution, but it worked. I envisioned a proxy that runs inside the Ruby world as a Rack server, so it could be mounted into our Rails application as a route. The browser visits a URL inside our domain and the WebSocket request just gets forwarded to the right endpoint.
+
+The inspiration for the Ruby implementation came from ActionCable. The Rack webserver exposes a socket hijacking interface that is able to expose the underlying socket of an HTTP request. The recipe is simple: if an HTTP request contains an "Upgrade: websocket" header, hijack the socket and return the worker thread for handling other requests. In the meantime initiate a connection to the other endpoint and when both sockets are available, pass them further to the proxy. In code it would look something like this:
+
+```ruby
+def call(env)
+  if env['HTTP_UPGRADE'] = 'websocket'
+    ws_sock = env['rack.hijack'].call
+    console_sock = SystemConsole.find_by!(:url_secret => env['REQUEST_URI'])
+
+    @proxy.push(ws_sock, console_sock)
+
+    return [200, {}, [""]]
+  else
+    return [500, {}, ["ERROR"]]
+  end
+end
+```
+
+The proxy itself needs to outlive a Rack request handling, however, it cannot run in a separate context as it needs to access the sockets. This can be easily achieved by using a  thread, even if threads are not the strongest side of Ruby due to the GVL. The job of the proxy is to transmit between one or more socket pairs with translation between WebSocket and regular TCP. A naive implementation could be to use a pair of threads for a pair of sockets with blocking read and write:
```

Comment:
What is GVL? Maybe expand it or link to an explanation. (as a non-Ruby programmer it's not something I understood right away)